temporalio/worker/_tuning.py

```diff
@@ -0,0 +1,185 @@
+    :py:class:`CompositeTuner`, all resource-based slot suppliers must use the same tuner options."""
+
+
+SlotSupplier: TypeAlias = Union[FixedSizeSlotSupplier, ResourceBasedSlotSupplier]
+
+
+def _to_bridge_slot_supplier(
+    slot_supplier: SlotSupplier, kind: Literal["workflow", "activity", "local_activity"]
+) -> temporalio.bridge.worker.SlotSupplier:
+    if isinstance(slot_supplier, FixedSizeSlotSupplier):
+        return temporalio.bridge.worker.FixedSizeSlotSupplier(slot_supplier.num_slots)
+    elif isinstance(slot_supplier, ResourceBasedSlotSupplier):
+        min_slots = 5 if kind == "workflow" else 1
+        max_slots = 500
+        ramp_throttle = (
+            timedelta(seconds=0) if kind == "workflow" else timedelta(milliseconds=50)
+        )
+        if slot_supplier.slot_options.minimum_slots is not None:
+            min_slots = slot_supplier.slot_options.minimum_slots
+        if slot_supplier.slot_options.maximum_slots is not None:
+            max_slots = slot_supplier.slot_options.maximum_slots
+        if slot_supplier.slot_options.ramp_throttle is not None:
+            ramp_throttle = slot_supplier.slot_options.ramp_throttle
+        return temporalio.bridge.worker.ResourceBasedSlotSupplier(
+            min_slots,
+            max_slots,
+            int(ramp_throttle / timedelta(milliseconds=1)),
+            temporalio.bridge.worker.ResourceBasedTunerOptions(
+                slot_supplier.tuner_options.target_memory_usage,
+                slot_supplier.tuner_options.target_cpu_usage,
+            ),
+        )
+    else:
+        raise TypeError(f"Unknown slot supplier type: {slot_supplier}")
+
+
+class WorkerTuner(ABC):
+    """WorkerTuners allow for the dynamic customization of some aspects of worker configuration"""
+
+    @abstractmethod
+    def _get_workflow_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_local_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+
+class ResourceBasedTuner(WorkerTuner):
+    """This tuner attempts to maintain certain levels of resource usage when under load.
+
+    .. warning::
+        The resource based tuner is currently experimental.
+    """
+
+    def __init__(self, options: ResourceBasedTunerOptions):
+        """Create a new ResourceBasedTuner
+
+        Args:
+            options: Specify the target resource usage levels for the tuner
+        """
+        self.options = options
+        self.workflow_task_options: Optional[ResourceBasedSlotOptions] = None
+        self.activity_task_options: Optional[ResourceBasedSlotOptions] = None
+        self.local_activity_task_options: Optional[ResourceBasedSlotOptions] = None
+
+    def set_workflow_task_options(self, options: ResourceBasedSlotOptions):
```

Comment:
I think these should be kwargs on the constructor and probably make this whole tuner a frozen dataclass. (EDIT: I have a better suggestion with a static method above)